File: backend/src/main/java/com/bigdata/portal/util/DistributeNews.java
package com.bigdata.portal.util;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class DistributeNews {
    
    private static final String[] TARGET_CHANNELS = {"AI", "大数据", "云计算", "互联网", "硬件", "创业"};
    
    private static final Map<String, List<String>> CHANNEL_KEYWORDS = new HashMap<>();
    static {
        List<String> aiKeywords = new ArrayList<>();
        aiKeywords.add("ai");
        aiKeywords.add("gpt");
        aiKeywords.add("llm");
        aiKeywords.add("大语言模型");
        aiKeywords.add("人工智能");
        aiKeywords.add("机器学习");
        aiKeywords.add("深度学习");
        aiKeywords.add("robot");
        aiKeywords.add("机器人");
        aiKeywords.add("算法");
        aiKeywords.add("模型");
        CHANNEL_KEYWORDS.put("AI", aiKeywords);
        
        List<String> bigDataKeywords = new ArrayList<>();
        bigDataKeywords.add("大数据");
        bigDataKeywords.add("数据");
        bigDataKeywords.add("spark");
        bigDataKeywords.add("hadoop");
        bigDataKeywords.add("数据库");
        bigDataKeywords.add("数据分析");
        CHANNEL_KEYWORDS.put("大数据", bigDataKeywords);
        
        List<String> cloudKeywords = new ArrayList<>();
        cloudKeywords.add("cloud");
        cloudKeywords.add("云");
        cloudKeywords.add("aws");
        cloudKeywords.add("azure");
        cloudKeywords.add("阿里云");
        cloudKeywords.add("云原生");
        cloudKeywords.add("docker");
        cloudKeywords.add("kubernetes");
        CHANNEL_KEYWORDS.put("云计算", cloudKeywords);
        
        List<String> internetKeywords = new ArrayList<>();
        internetKeywords.add("互联网");
        internetKeywords.add("web");
        internetKeywords.add("app");
        internetKeywords.add("应用");
        internetKeywords.add("平台");
        internetKeywords.add("产品");
        internetKeywords.add("运营");
        CHANNEL_KEYWORDS.put("互联网", internetKeywords);
        
        List<String> hardwareKeywords = new ArrayList<>();
        hardwareKeywords.add("硬件");
        hardwareKeywords.add("芯片");
        hardwareKeywords.add("cpu");
        hardwareKeywords.add("gpu");
        hardwareKeywords.add("手机");
        hardwareKeywords.add("iphone");
        hardwareKeywords.add("pc");
        hardwareKeywords.add("电脑");
        hardwareKeywords.add("设备");
        hardwareKeywords.add("半导体");
        CHANNEL_KEYWORDS.put("硬件", hardwareKeywords);
        
        List<String> startupKeywords = new ArrayList<>();
        startupKeywords.add("创业");
        startupKeywords.add("投资");
        startupKeywords.add("融资");
        startupKeywords.add("公司");
        startupKeywords.add("企业");
        startupKeywords.add("techcrunch");
        startupKeywords.add("创业公司");
        CHANNEL_KEYWORDS.put("创业", startupKeywords);
    }
    
    public static void main(String[] args) {
        String jdbcUrl = "jdbc:mysql://192.168.146.128:3306/bigdata_portal?useSSL=false&characterEncoding=utf8&serverTimezone=Asia/Shanghai&allowPublicKeyRetrieval=true";
        String username = "root";
        String password = "root123";
        
        Random random = new Random();
        
        System.out.println("======================================================================");
        System.out.println("重新分配新闻到各个频道");
        System.out.println("======================================================================");
        
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {
            
            // 先查看当前频道分布
            System.out.println("\n当前频道分布:");
            ResultSet rs = stmt.executeQuery("SELECT channel, COUNT(*) as count FROM news GROUP BY channel");
            while (rs.next()) {
                System.out.println("  " + rs.getString("channel") + ": " + rs.getInt("count") + " 条");
            }
            rs.close();
            
            // 获取所有新闻
            System.out.println("\n获取所有新闻...");
            List<Map<String, Object>> allNews = new ArrayList<>();
            rs = stmt.executeQuery("SELECT id, title, summary FROM news");
            while (rs.next()) {
                Map<String, Object> news = new HashMap<>();
                news.put("id", rs.getLong("id"));
                news.put("title", rs.getString("title"));
                news.put("summary", rs.getString("summary"));
                allNews.add(news);
            }
            rs.close();
            
            System.out.println("获取到 " + allNews.size() + " 条新闻");
            
            // 开始重新分配
            System.out.println("\n开始重新分配...");
            int updatedCount = 0;
            
            for (Map<String, Object> news : allNews) {
                Long newsId = (Long) news.get("id");
                String title = ((String) news.getOrDefault("title", "")).toLowerCase();
                String summary = ((String) news.getOrDefault("summary", "")).toLowerCase();
                
                // 根据关键词分配
                String newChannel = null;
                for (Map.Entry<String, List<String>> entry : CHANNEL_KEYWORDS.entrySet()) {
                    String channel = entry.getKey();
                    List<String> keywords = entry.getValue();
                    for (String keyword : keywords) {
                        if (title.contains(keyword) || summary.contains(keyword)) {
                            newChannel = channel;
                            break;
                        }
                    }
                    if (newChannel != null) break;
                }
                
                // 如果没有匹配到，随机分配
                if (newChannel == null) {
                    newChannel = TARGET_CHANNELS[random.nextInt(TARGET_CHANNELS.length)];
                }
                
                // 更新数据库
                String sql = "UPDATE news SET channel = '" + newChannel + "' WHERE id = " + newsId;
                stmt.executeUpdate(sql);
                updatedCount++;
                
                if (updatedCount % 10 == 0) {
                    System.out.println("  已更新 " + updatedCount + " 条...");
                }
            }
            
            System.out.println("\n✓ 成功重新分配 " + updatedCount + " 条新闻!");
            
            // 查看新的频道分布
            System.out.println("\n新的频道分布:");
            rs = stmt.executeQuery("SELECT channel, COUNT(*) as count FROM news GROUP BY channel");
            while (rs.next()) {
                String channel = rs.getString("channel");
                int count = rs.getInt("count");
                String status = count > 0 ? "✓" : "✗";
                System.out.println("  " + status + " " + channel + ": " + count + " 条");
            }
            rs.close();
            
            System.out.println("\n======================================================================");
            System.out.println("分配完成!");
            System.out.println("======================================================================");
            
        } catch (Exception e) {
            System.err.println("\n✗ 错误: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: distribute_news.py
import requests
import random
import pymysql

base_url = "http://192.168.146.128"
db_config = {
    'host': '192.168.146.128',
    'user': 'root',
    'password': 'root123',
    'database': 'bigdata_portal',
    'port': 3306
}

print("=" * 70)
print("重新分配新闻到各个频道")
print("=" * 70)

# 获取所有新闻
print("\n获取现有新闻...")
response = requests.get(f"{base_url}/api/news")
all_news = []
if response.status_code == 200:
    all_news = response.json().get('data', [])
    print(f"获取到 {len(all_news)} 条新闻")

if not all_news:
    print("没有获取到新闻，退出")
    exit(1)

# 定义目标频道和分配比例
target_channels = ['AI', '大数据', '云计算', '互联网', '硬件', '创业']
channel_keywords = {
    'AI': ['ai', 'gpt', 'llm', '大语言模型', '人工智能', '机器学习', '深度学习', 'robot', '机器人', '算法', '模型'],
    '大数据': ['big data', '大数据', '数据', 'spark', 'hadoop', '数据库', '数据分析'],
    '云计算': ['cloud', '云', 'aws', 'azure', 'aliyun', '阿里云', '云原生', 'docker', 'kubernetes'],
    '互联网': ['互联网', 'web', 'app', '应用', '平台', '产品', '运营', '36氪', '虎嗅', '钛媒体', '极客公园'],
    '硬件': ['硬件', '芯片', 'cpu', 'gpu', '手机', 'iphone', 'pc', '电脑', '设备', '爱范儿', '半导体'],
    '创业': ['创业', '投资', '融资', '公司', '企业', 'techcrunch', '创业公司'],
}

# 连接数据库
print("\n连接数据库...")
conn = None
try:
    conn = pymysql.connect(**db_config)
    cursor = conn.cursor()
    
    # 先查看当前频道分布
    print("\n当前频道分布:")
    cursor.execute("SELECT channel, COUNT(*) as count FROM news GROUP BY channel")
    for row in cursor:
        print(f"  {row[0]}: {row[1]} 条")
    
    # 开始重新分配
    print(f"\n开始重新分配 {len(all_news)} 条新闻...")
    updated_count = 0
    
    for news in all_news:
        news_id = news.get('id')
        title = news.get('title', '').lower()
        summary = news.get('summary', '').lower()
        
        # 根据关键词分配
        new_channel = None
        for channel, keywords in channel_keywords.items():
            for keyword in keywords:
                if keyword in title or keyword in summary:
                    new_channel = channel
                    break
            if new_channel:
                break
        
        # 如果没有匹配到，随机分配
        if not new_channel:
            new_channel = random.choice(target_channels)
        
        # 更新数据库
        sql = "UPDATE news SET channel = %s WHERE id = %s"
        cursor.execute(sql, (new_channel, news_id))
        updated_count += 1
        
        if updated_count % 10 == 0:
            print(f"  已更新 {updated_count} 条...")
    
    conn.commit()
    print(f"\n✓ 成功重新分配 {updated_count} 条新闻!")
    
    # 查看新的频道分布
    print("\n新的频道分布:")
    cursor.execute("SELECT channel, COUNT(*) as count FROM news GROUP BY channel")
    for row in cursor:
        count = row[1]
        status = "✓" if count > 0 else "✗"
        print(f"  {status} {row[0]}: {count} 条")
    
    print("\n" + "=" * 70)
    print("分配完成!")
    print("=" * 70)
    
except Exception as e:
    print(f"\n✗ 错误: {e}")
    if conn:
        conn.rollback()
finally:
    if conn:
        conn.close()
